kratos/python_scripts/sympy_fe_utilities.py

```diff
@@ -1,466 +1,522 @@
-            newstring = outstring.replace(to_be_replaced, replacement_string)
-            outstring = newstring
+        _ReplaceIndices(language, outstring)
 
     return outstring
 
-def OutputMatrix_CollectingFactors(A, name, mode, initial_tabs=3, max_index=30, optimizations='basic', replace_indices=True, assignment_op="="):
-    """ This method collects the constants of the replacement for matrices
+def _OutputX_CollectionFactors(A, name, language, indentation_level, optimizations, replace_indices, assignment_op, output_func):
+    """
+    This method collects the constants of the replacement for matrices, vectors and scalars.
 
     Keyword arguments:
-    A -- The  factors
-    name -- The name of the constant
-    mode -- The mode of replacement
-    initial_tabs -- The number of initial tabulations
-    max_index -- The max number of indexes
-    optimizations -- The level of optimizations
-    replace_indices -- If the indixes must be replaced
-    assignment_op -- The assignment operation
-    """
-    symbol_name = "c"+name
-    A_factors, A_collected = cse(A,numbered_symbols(symbol_name), optimizations)
+    - A -- The  factors
+    - name -- The name of the constant
+    - language -- The language of replacement
+    - indentation_level -- The depth of the indentation (4 spaces per level)
+    - optimizations -- The level of optimizations
+    - replace_indices -- Set to `True` to replace matrix[i,j] with matrix_i_j (And similarly for vectors)
+    - assignment_op -- The assignment operation
+    - output_func -- The output function. Must have the same signature as OutputMatrix and OutputVector
+    """
+    symbol_name = "c" + name
+    A_factors, A_collected = sympy.cse(A, sympy.numbered_symbols(symbol_name), optimizations)
     A = A_collected[0] #overwrite lhs with the one with the collected components
 
     Acoefficient_str = str("")
     for factor in A_factors:
-        varname = factor[0]
+        varname = str(factor[0])
         value = factor[1]
-        output_value = OutputSymbolicVariable(value, mode,initial_tabs, max_index, replace_indices)
-        Acoefficient_str += "const double " + str(varname.__str__()) + " = " + output_value
-        #print(output_str)
-    A_out = Acoefficient_str + OutputMatrix(A, name, mode, initial_tabs, max_index, replace_indices, assignment_op)
+        Acoefficient_str += OutputSymbolicVariableDeclaration(value, varname, language, indentation_level, replace_indices)
+
+    A_out = Acoefficient_str + output_func(A, name, language, indentation_level, replace_indices, assignment_op)
     return A_out
 
-def OutputVector_CollectingFactors(A, name, mode, initial_tabs=3, max_index=30, optimizations='basic', replace_indices=True, assignment_op="="):
-    """ This method collects the constants of the replacement for vectors
+
+def OutputMatrix_CollectingFactors(A, name, language, indentation_level=0, max_index=None, optimizations='basic', replace_indices=True, assignment_op="="):
```

Comment:
We should eventually think about refactoring `OutputMatrix_CollectingFactors` to our standard (`OutputMatrixCollectingFactors`).

Reply:
I agree. I prefered making as few breaking changes as possible because symbolic generation is not well tested and an issue can remain burried for a long time.